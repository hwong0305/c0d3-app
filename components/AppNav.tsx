import React, { useEffect } from 'react'
import NavLink from './NavLink'
import Button from './Button'
import { useMutation } from '@apollo/react-hooks'
import { LOGOUT_USER } from '../graphql/queries'
import _ from 'lodash'

import '../scss/navbar.scss'

type AuthButtonProps = {
  initial: string
  username: string
}

type Props = {
  loggedIn?: boolean
  username?: string
  firstName?: string
  lastName?: string
}

const AuthLink = () => (
  <div className="navbar-nav collapse navbar-collapse">
    <NavLink
      path="/curriculum"
      activePath="/curriculum"
      className="nav-item nav-link"
    >
      Curriculum
    </NavLink>
    <NavLink path="#" className="nav-item nav-link">
      Repo
    </NavLink>
    <NavLink path="#" className="nav-item nav-link">
      Journey
    </NavLink>
    <NavLink
      path="https://chat.c0d3.com"
      className="nav-item nav-link"
      external
    >
      Help
    </NavLink>
  </div>
)

const AuthButton: React.FC<AuthButtonProps> = ({ initial, username }) => {
  const [logoutUser, { data }] = useMutation(LOGOUT_USER)
  useEffect(() => {
    const { success } = _.get(data, 'logout', false)
    if (success) {
      window.location.pathname = '/'
    }
  }, [data])
  return (
    <div>
      <Button
        btnType="border btn-secondary overflow-hidden p-2 text-truncate"
        initial={initial}
        text={username}
      />
      <Button
        text="Logout"
        btnType="border btn-secondary ml-2"
        onClick={async () => {
          await logoutUser()
        }}
      />
    </div>
  )
}

const UnAuthButton = () => (
  <div>
    <NavLink path="/login" className="btn btn-secondary border mr-3">
      Login
    </NavLink>
    <NavLink path="/signup" className="btn btn-secondary border mr-3">
      Signup
    </NavLink>
  </div>
)

const UnAuthLink = () => (
  <div className="navbar-nav collapse navbar-collapse">
    <NavLink path="/" activePath="/" className="nav-item nav-link">
      Home
    </NavLink>
    <NavLink path="/#learning" className="nav-item nav-link">
      Learning Process
    </NavLink>
    <NavLink
      path="https://c0d3.com/book"
      className="nav-item nav-link"
      external
    >
      Resources
    </NavLink>
    <NavLink
      path="https://chat.c0d3.com"
      className="nav-item nav-link"
      external
    >
      Help
    </NavLink>
  </div>
)

const AppNav: React.FC<Props> = ({
  loggedIn = false,
  username = '',
  firstName,
  lastName
}) => {
  const initial = firstName && lastName ? firstName[0] + lastName[0] : ''
  return (
    <nav className="navbar navbar-expand-lg navbar-light justify-content-between bg-white">
      <div className="container">
        <NavLink
          path="/"
          className="navbar-brand text-primary font-weight-bold"
        >
          C0D3
        </NavLink>
        <div id="navbarNav">
          <div className="navbar-nav collapse navbar-collapse">
            {loggedIn ? <AuthLink /> : <UnAuthLink />}
          </div>
        </div>
        {loggedIn ? (
          <AuthButton initial={initial} username={username} />
        ) : (
          <UnAuthButton />
        )}
      </div>
    </nav>
  )
}

export default AppNav
