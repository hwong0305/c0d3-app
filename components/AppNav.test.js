jest.mock('../helpers/logoutUser')
jest.mock('next/router')
import React from 'react'
import { render, fireEvent, wait } from '@testing-library/react'
import AppNav from './AppNav'
import logoutUser from '../helpers/logoutUser'

// Mock global.window
global.window = Object.create(window)

// define property to be modifed beforeEach test
Object.defineProperty(global.window, 'location', {
  value: { pathname: '/curriculum' } // make sure pathname isnt '/' by default
})

describe('AppNav Component', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  test('Should match snapshot with default params', () => {
    const { container } = render(<AppNav />)
    expect(container).toMatchSnapshot()
  })

  test('Should match snapshot when loggedIn', () => {
    const { container } = render(<AppNav loggedIn />)
    expect(container).toMatchSnapshot()
  })

  test('Should call logoutUser when logout Button clicked', () => {
    const { getByText } = render(<AppNav loggedIn />)
    const leftClick = { button: 0 }

    expect(logoutUser).not.toHaveBeenCalled()
    fireEvent.click(getByText('Logout'), leftClick)
    expect(logoutUser).toHaveBeenCalledTimes(1)
  })

  test('Should redirect to / route when logout success', async () => {
    const { getByText } = render(<AppNav loggedIn />)
    const leftClick = { button: 0 }

    logoutUser.mockReturnValue(Promise.resolve(true))
    fireEvent.click(getByText('Logout'), leftClick)
    await wait(() => expect(global.window.location.pathname).toEqual('/'))
  })
})
