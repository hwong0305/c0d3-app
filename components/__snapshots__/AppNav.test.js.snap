// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppNav Component Should match snapshot when loggedIn 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link active"
              href="/curriculum"
            >
              <span
                class="sr-only"
              >
                (current)
              </span>
              Curriculum
            </a>
            <a
              class="nav-item nav-link"
              href="#"
            >
              Repo
            </a>
            <a
              class="nav-item nav-link"
              href="#"
            >
              Journey
            </a>
            <a
              class="nav-item nav-link"
              href="https://chat.c0d3.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              Help
            </a>
          </div>
        </div>
      </div>
      <div>
        <button
          class="btn border btn-secondary overflow-hidden p-2 text-truncate"
        />
        <button
          class="btn border btn-secondary ml-2"
        >
          Logout
        </button>
      </div>
    </div>
  </nav>
</div>
`;

exports[`AppNav Component Should match snapshot with default params 1`] = `
<div>
  <nav
    class="navbar navbar-expand-lg navbar-light justify-content-between bg-white"
  >
    <div
      class="container"
    >
      <a
        class="navbar-brand text-primary font-weight-bold"
        href="/"
      >
        C0D3
      </a>
      <div
        id="navbarNav"
      >
        <div
          class="navbar-nav collapse navbar-collapse"
        >
          <div
            class="navbar-nav collapse navbar-collapse"
          >
            <a
              class="nav-item nav-link active"
              href="/"
            >
              <span
                class="sr-only"
              >
                (current)
              </span>
              Home
            </a>
            <a
              class="nav-item nav-link"
              href="/#learning"
            >
              Learning Process
            </a>
            <a
              class="nav-item nav-link"
              href="https://c0d3.com/book"
              rel="noopener noreferrer"
              target="_blank"
            >
              Resources
            </a>
            <a
              class="nav-item nav-link"
              href="https://chat.c0d3.com"
              rel="noopener noreferrer"
              target="_blank"
            >
              Help
            </a>
          </div>
        </div>
      </div>
      <div>
        <a
          class="btn btn-secondary border mr-3"
          href="/login"
        >
          Login
        </a>
        <a
          class="btn btn-secondary border mr-3"
          href="/signup"
        >
          Signup
        </a>
      </div>
    </div>
  </nav>
</div>
`;
